plugins/module_utils/copy.py

```diff
@@ -216,6 +228,87 @@ def copy_vsam_ps(src, dest):
     return rc, out, err
 
 
+def copy_asa_uss2mvs(src, dest):
+    """Copy a file from USS to an ASA sequential data set or PDS/E member.
+
+    Arguments:
+        src: {str} -- Path of the USS file
+        dest: {str} -- The MVS destination data set or member
+
+    Returns:
+        boolean -- The return code after the copy command executed successfully
+        str -- The stdout after the copy command executed successfully
+        str -- The stderr after the copy command executed successfully
+    """
+    module = AnsibleModuleHelper(argument_spec={})
+
+    oget_cmd = 'tsocmd "OGET \'{0}\' \'{1}\'"'.format(src, dest)
+    rc, out, err = module.run_command(oget_cmd)
+
+    return TSOCmdResponse(rc, out, err)
+
+
+def copy_asa_mvs2uss(src, dest):
+    """Copy an ASA sequential data set or member to USS.
+
+    Arguments:
+        src: {str} -- The MVS data set to be copied
+        dest: {str} -- Destination path in USS
+
+    Returns:
+        boolean -- The return code after the copy command executed successfully
+        str -- The stdout after the copy command executed successfully
+        str -- The stderr after the copy command executed successfully
+    """
+    module = AnsibleModuleHelper(argument_spec={})
+    src = _validate_data_set_name(src)
+    dest = _validate_path(dest)
+
+    oput_cmd = 'tsocmd "OPUT \'{0}\' \'{1}\'"'.format(src, dest)
+    rc, out, err = module.run_command(oput_cmd)
+
+    return TSOCmdResponse(rc, out, err)
+
+
+def copy_asa_pds2uss(src, dest):
+    """Copy all members from an ASA PDS/E to USS.
+
+    Arguments:
+        src: {str} -- The MVS data set to be copied
+        dest: {str} -- Destination path in USS (must be a directory)
+
+    Returns:
+        boolean -- The return code after the copy command executed successfully
+        str -- The stdout after the copy command executed successfully
+        str -- The stderr after the copy command executed successfully
+    """
+    from os import path
+    from zoautil_py import datasets
+
+    module = AnsibleModuleHelper(argument_spec={})
+    src = _validate_data_set_name(src)
+    dest = _validate_path(dest)
+
+    for member in datasets.list_members(src):
```

Comment:
When iterating and encountering a non-zero RC and failing, do we leave a partially or any PDSE on the managed z/OS node? I recall we would have some clean up in a module failure but did not follow the code all the way, maybe you know off hand.

Reply:
We leave an incomplete PDSE in case of failure, we removed the code that would try to clean up because it was giving us more problems than it fixed

Reply:
I see the complexity that can come with this so I opened a research issue, I don't think it should be further addressed at this point , thank you for the update, here is the issue. https://github.com/ansible-collections/ibm_zos_core/issues/1033